src/Microsoft.Health.Fhir.Shared.Api/Controllers/ValidateController.cs

```diff
@@ -39,23 +42,58 @@ public ValidateController(IMediator mediator)
         [HttpPost]
         [Route(KnownRoutes.ValidateResourceType)]
         [AuditEventType(AuditEventSubType.Read)]
-        public async Task<IActionResult> Validate([FromBody] Resource resource)
+        public async Task<IActionResult> Validate([FromBody] Resource resource, [FromQuery(Name = "profile")] Uri profile)
         {
-            return await RunValidationAsync(resource);
+            if (resource.ResourceType == ResourceType.Parameters)
+            {
+                var parameterResource = (Parameters)resource;
+                var profileFromParameters = parameterResource.Parameter.Find(param => param.Name.Equals("profile", StringComparison.OrdinalIgnoreCase));
+                if (profileFromParameters != null)
+                {
+                    if (profile != null)
+                    {
+                        throw new BadRequestException(Api.Resources.MultipleProfilesProvided);
+                    }
+
+                    if (profileFromParameters.Value == null)
+                    {
+                        throw new BadRequestException("Profile parameter is empty");
+                    }
+
+                    try
+                    {
+                        profile = new Uri(profileFromParameters.Value.ToString());
+                    }
+                    catch
+                    {
+                        throw new BadRequestException("Profile is invalid.");
+                    }
+                }
+
+                resource = parameterResource.Parameter.Find(param => param.Name.Equals("resource", StringComparison.OrdinalIgnoreCase)).Resource;
+            }
+
+            return await RunValidationAsync(resource, profile);
         }
 
-        [HttpPost]
+        [HttpGet]
         [Route(KnownRoutes.ValidateResourceTypeById)]
         [AuditEventType(AuditEventSubType.Read)]
-        [ValidateResourceIdFilter(true)]
-        public async Task<IActionResult> ValidateById([FromBody] Resource resource)
+        public async Task<IActionResult> ValidateById([FromRoute] string typeParameter, [FromRoute] string idParameter, [FromQuery] Uri profile)
         {
-            return await RunValidationAsync(resource);
+            // Read resource from storage.
+            RawResourceElement response = await _mediator.GetResourceAsync(new ResourceKey(typeParameter, idParameter), HttpContext.RequestAborted);
+
+            // Convert it to fhir object.
+            FhirJsonParser parser = new FhirJsonParser();
+            var convertedResource = parser.Parse<Resource>(response.RawResource.Data);
```

Comment:
The interface might not be exposed, but this class should handle this
https://github.com/microsoft/fhir-server/blob/main/src/Microsoft.Health.Fhir.Core/Features/Persistence/ResourceDeserializer.cs#L14

Reply:
I knew this functionality should exist, I just couldn't find it! Thanks!

---
In reply to: [574164155](https://github.com/microsoft/fhir-server/pull/1654#discussion_r574164155) [](ancestors = 574164155)